kaldb/src/test/java/com/slack/kaldb/clusterManager/RecoveryTaskAssignmentServiceTest.java

```diff
@@ -0,0 +1,724 @@
+import com.slack.kaldb.metadata.recovery.RecoveryTaskMetadata;
+import com.slack.kaldb.metadata.recovery.RecoveryTaskMetadataStore;
+import com.slack.kaldb.metadata.zookeeper.MetadataStore;
+import com.slack.kaldb.metadata.zookeeper.ZookeeperMetadataStoreImpl;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.slack.kaldb.proto.metadata.Metadata;
+import com.slack.kaldb.testlib.MetricsUtil;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.curator.test.TestingServer;
+import org.assertj.core.api.Assertions;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class RecoveryTaskAssignmentServiceTest {
+
+  private TestingServer testingServer;
+  private MeterRegistry meterRegistry;
+
+  private MetadataStore metadataStore;
+  private RecoveryTaskMetadataStore recoveryTaskMetadataStore;
+  private RecoveryNodeMetadataStore recoveryNodeMetadataStore;
+
+  @Before
+  public void setup() throws Exception {
+    Tracing.newBuilder().build();
+    meterRegistry = new SimpleMeterRegistry();
+    testingServer = new TestingServer();
+
+    KaldbConfigs.ZookeeperConfig zkConfig =
+        KaldbConfigs.ZookeeperConfig.newBuilder()
+            .setZkConnectString(testingServer.getConnectString())
+            .setZkPathPrefix("RecoveryTaskAssignmentServiceTest")
+            .setZkSessionTimeoutMs(1000)
+            .setZkConnectionTimeoutMs(1000)
+            .setSleepBetweenRetriesMs(1000)
+            .build();
+
+    metadataStore = ZookeeperMetadataStoreImpl.fromConfig(meterRegistry, zkConfig);
+    recoveryTaskMetadataStore = spy(new RecoveryTaskMetadataStore(metadataStore, true));
+    recoveryNodeMetadataStore = spy(new RecoveryNodeMetadataStore(metadataStore, true));
+  }
+
+  @After
+  public void shutdown() throws IOException {
+    recoveryNodeMetadataStore.close();
+    recoveryTaskMetadataStore.close();
+    metadataStore.close();
+
+    testingServer.close();
+    meterRegistry.close();
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void shouldCheckInvalidConfigs() {
+    KaldbConfigs.ManagerConfig.RecoveryTaskAssignmentServiceConfig
+        recoveryTaskAssignmentServiceConfig =
+            KaldbConfigs.ManagerConfig.RecoveryTaskAssignmentServiceConfig.newBuilder()
+                .setEventAggregationSecs(-1)
+                .setScheduleInitialDelayMins(1)
+                .setSchedulePeriodMins(10)
+                .build();
+
+    new RecoveryTaskAssignmentService(
+        recoveryTaskMetadataStore,
+        recoveryNodeMetadataStore,
+        recoveryTaskAssignmentServiceConfig,
+        meterRegistry);
+  }
+
+  @Test
+  public void shouldHandleNoNodesOrTasks() {
```

Comment:
I think its also good to add some simple unit tests when there is only 1 node and 1 task to be assigned. One node is assigned and one node is free and a task is assigned to it.

Also, most of these unit tests manually run the assignment operation. It would be good to have these tests test that the automatic execution of the method here.

Currently, the notifications are also fired when a node data changes. It would be good to add some tests to make sure that on node data changes the assignment operations terminate early.

Reply:
All of the tests here are unit tests except for the two lifecycle tests (`shouldHandleTasksAvailableFirstLifecycle`, `shouldHandleNodesAvailableFirstLifecycle`). In both of those we do as you suggested with properly bootstrapping the service as a normal Guava service and ensure that the assignments are working as expected, and then ensures that the service properly can shutdown.